dan_layer/consensus/src/hotstuff/on_receive_proposal.rs

```diff
@@ -218,16 +348,60 @@ where TConsensusSpec: ConsensusSpec
         Ok(())
     }
 
-    async fn handle_foreign_proposal(&self, from: TConsensusSpec::Addr, block: Block) -> Result<(), HotStuffError> {
-        let vn = self.epoch_manager.get_validator_node(block.epoch(), &from).await?;
+    async fn handle_foreign_proposal(
+        &self,
+        from: TConsensusSpec::Addr,
+        local_committee: &Committee<TConsensusSpec::Addr>,
+        block: Block,
+    ) -> Result<(), HotStuffError> {
+        if self.store.with_read_tx(|tx| block.justify().exists(tx))? {
+            // We already seen this block. And the block we saw was valid.
+            return Ok(());
+        }
+
         let committee_shard = self
             .epoch_manager
-            .get_committee_shard(block.epoch(), vn.shard_key)
+            .get_committee_shard(block.epoch(), *block.proposed_by())
             .await?;
         self.validate_proposed_block(&from, &block)?;
         self.store
             .with_write_tx(|tx| self.on_receive_foreign_block(tx, &block, &committee_shard))?;
 
+        // If we received the foreign proposal, we send it to the leader (if we are not the leader), the leader then
+        // redistributes the block to all other nodes. This way if the leader is not faulty O(n) messages will be send
+        // around. If the leader doesn't have the message it will take 2 delta (if the delta time is the maximum latency
+        // between any two nodes) to have it everywhere. If the leader has the message already it will be just 1 delta.
+        // Worst case scenario is when we have f faulty nodes, and 2f honest nodes have the message and 1 node doesnt,
+        // but he is the (f+1)th leader. In this case we send exactly 2f*f+2f+3f messages around. 2f*f to the
+        // faulty leaders, 2f to the honest leader, and 3f from the leader.
+        if self
+            .leader_strategy
+            .is_leader(&self.validator_addr, local_committee, block.id(), block.height())
```

Comment:
This block will not be in our chain, so this call will not return the correct leader. You might have to get the `last_voted` block or highest block in our local chain to determine who our current leader is

Reply:
Done. I had to also change other methods so I can send the leader around when I want to send the foreign block.